File: 207CourseSchedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        
        # store the prereqs in a hashmap
        # dfs through the graph
        # have a visited set to find out circular loop... thats the invalid case.
        preMap = {i:[] for i in range(numCourses)}
        
        for crs, pre in prerequisites:
            preMap[crs].append(pre)
        
        visitedSet = set()
        
        def dfs(crs):
            if crs in visitedSet:
                return False
            if preMap[crs] == []:
                return True
            
            visitedSet.add(crs)
            for pre in preMap[crs]:
                if not dfs(pre):
                    return False
            visitedSet.remove(crs)
            preMap[crs] = []
            return True
        
        for crs in range(numCourses):
            if not dfs(crs):
                return False
            
        return True

File: 207.CourseSchedule.go
func canFinish(numCourses int, prerequisites [][]int) bool {
    preMap := make(map[int][]int)

    for _,prerq := range prerequisites{
        preMap[prerq[1]] =  append(preMap[prerq[1]], prerq[0])
    }

    visit := make(map[int]struct{})

    var dfs func(int) bool
    dfs = func( crs int) bool{
        if _,ok := visit[crs]; ok{
            return false
        }

        if len(preMap[crs]) == 0{
            return true
        }

        visit[crs] = struct{}{}

        for _,pre := range preMap[crs]{
            if !dfs(pre){
                return false
            }
        }

        delete(visit, crs)
        preMap[crs] = []int{}
        return true
    }

    for i := 0; i < numCourses ; i++ {
        if !dfs(i){
            return false
        }
    }
    return true
}
